document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java

```diff
@@ -0,0 +1,347 @@
+package org.hypertrace.core.documentstore.postgres;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.annotations.VisibleForTesting;
+import org.hypertrace.core.documentstore.Collection;
+import org.hypertrace.core.documentstore.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.sql.*;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class PostgresCollection implements Collection {
+  
+  private static final String ID_KEY = "id";
+  private static final String DOCUMENT_KEY = "document";
+  private static final Logger LOGGER = LoggerFactory.getLogger(PostgresCollection.class);
+  private static final String UPDATED_AT = "updated_at";
+  private static final String CREATED_AT = "created_at";
+  private static final Set<String> OUTER_COLUMNS = new HashSet<>() {{
+    add(CREATED_AT);
+    add(ID_KEY);
+    add(UPDATED_AT);
+  }};
+  
+  private Connection client;
+  private String collectionName;
+  private String columnType;
+  
+  public PostgresCollection(Connection client, String collectionName, String columnType) {
+    this.client = client;
+    this.collectionName = collectionName;
+    this.columnType = columnType;
+  }
+  
+  @Override
+  public boolean upsert(Key key, Document document) {
+    String jsonString = document.toJson();
+    String upsertSQL = String.format("INSERT INTO %s (%s,%s) VALUES( ?, ? :: %s) ON CONFLICT(%s) DO UPDATE SET %s = ?::%s ",
+      collectionName, ID_KEY, DOCUMENT_KEY, columnType, ID_KEY, DOCUMENT_KEY, columnType);
+    try {
+      
+      PreparedStatement preparedStatement = client.prepareStatement(upsertSQL, Statement.RETURN_GENERATED_KEYS);
+      preparedStatement.setString(1, key.toString());
+      preparedStatement.setString(2, jsonString);
+      preparedStatement.setString(3, jsonString);
+      ResultSet resultSet = preparedStatement.executeQuery();
+      
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug("Write result: " + resultSet.toString());
+      }
+      
+      return resultSet.rowUpdated();
+    } catch (SQLException e) {
+      LOGGER.error("SQLException inserting document. key: {} content:{}", key, document, e);
+    }
+    return false;
+  }
+  
+  @Override
+  public boolean updateSubDoc(Key key, String subDocPath, Document subDocument) {
+    return false;
+  }
+  
+  @Override
+  public Iterator<Document> search(Query query) {
```

Comment:
Currently, there is an assumption in HT code that each document will have an _id field.
e.g https://github.com/hypertrace/entity-service/blob/572b6eaa4db0b898967258faa7f3f12f43d122cc/entity-service-impl/src/main/java/org/hypertrace/entity/data/service/EntityDataServiceImpl.java#L533 
we can see that the usage of constant `EntityServiceConstants.ID` in the above method. It is added to the query filter. We need to create one more issue in HT to fix this.

But, for now, I would suggest if we can add the _id field in each document in the `upsert` and `bulkupsert` method.
```
String jsonString = document.toJson();
// convert to JSON, and we can add root level field in document
{
_id : key.toString()
// rest of document document.toJson()
}
```